core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -0,0 +1,533 @@
+    }
+
+    @VisibleForTesting
+    BinPackingNodeAllocatorService(NodeScheduler nodeScheduler, int maxTasksPerNode, int maxFullNodesPerQuery)
+    {
+        this.nodeScheduler = requireNonNull(nodeScheduler, "nodeScheduler is null");
+        this.maxTasksPerNode = maxTasksPerNode;
+        this.maxFullNodesPerQuery = maxFullNodesPerQuery;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (started.compareAndSet(false, true)) {
+            executor.schedule(() -> {
+                while (!stopped.get()) {
+                    try {
+                        processSemaphore.tryAcquire(PROCESS_PENDING_ACQUIRES_DELAY_SECONDS, TimeUnit.SECONDS);
+                        processSemaphore.drainPermits();
+                        processPendingAcquires();
+                    }
+                    catch (InterruptedException e) {
+                        currentThread().interrupt();
+                    }
+                    catch (Exception e) {
+                        // ignore to avoid getting unscheduled
+                        log.warn(e, "Error updating nodes");
+                    }
+                }
+            }, 0, TimeUnit.SECONDS);
+        }
+    }
+
+    @VisibleForTesting
+    void wakeupProcessPendingAcquires()
+    {
+        processSemaphore.release();
+    }
+
+    @VisibleForTesting
+    synchronized void processPendingAcquires()
+    {
+        processFullNodePendingAcquires();
+        processSharedPendingAcquires();
+    }
+
+    private void processSharedPendingAcquires()
+    {
+        Map<PendingAcquire, InternalNode> assignedNodes = new IdentityHashMap<>();
+        Map<PendingAcquire, RuntimeException> failures = new IdentityHashMap<>();
+        synchronized (this) {
+            Iterator<PendingAcquire> iterator = sharedPendingAcquires.iterator();
+            while (iterator.hasNext()) {
+                PendingAcquire pendingAcquire = iterator.next();
+                if (pendingAcquire.getFuture().isCancelled()) {
+                    iterator.remove();
+                    continue;
+                }
+                try {
+                    List<InternalNode> candidates = selectCandidates(pendingAcquire.getNodeRequirements(), pendingAcquire.getNodeSelector());
+                    if (candidates.isEmpty()) {
+                        throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+                    }
+
+                    Optional<InternalNode> node = tryAcquireSharedNode(candidates);
+                    if (node.isPresent()) {
+                        iterator.remove();
+                        assignedNodes.put(pendingAcquire, node.get());
+                    }
+                }
+                catch (RuntimeException e) {
+                    iterator.remove();
+                    failures.put(pendingAcquire, e);
+                }
+            }
+        }
+
+        // complete futures outside of synchronized section
```

Comment:
This method is called from `processPendingAcquires` which is `synchronized`. You should probably add an assert at the beginning of this method to verify that the method is not called when the lock is held. (same for the `processFullNodePendingAcquires`)

Reply:
good catch. `processPendingAcquires` should not be synchronized - dropped.